### PR TITLE
[THREESCALE-1980] pagination for application plan limits

### PR DESCRIPTION
**Pull request by macejmic**
**What this PR does / why we need it**:

Page renders slow so a pagination has to be added.

**Which issue(s) this PR fixes** 

https://issues.jboss.org/browse/THREESCALE-1980

### CODE REVIEW

File: app/controllers/admin/api/application_plan_limits_controller.rb [hunk 1]
```diff
@@ -14,15 +14,19 @@ class Admin::Api::ApplicationPlanLimitsController < Admin::Api::BaseController
...
+  #
   ##~ op.parameters.add @parameter_access_token
+  ##~ op.parameters.add @parameter_page
+  ##~ op.parameters.add @parameter_per_page_50
   ##~ op.parameters.add @parameter_application_plan_id_by_id_name
   #
-
   def index
-    respond_with(usage_limits)
+    respond_with(usage_limits.paginate(page: current_page, per_page: per_page))
```
Comment (josemigallas) [inline app/controllers/admin/api/application_plan_limits_controller.rb]: Just for consistency, I'd use the same names
```suggestion
    respond_with(usage_limits.paginate(current_page: current_page, per_page: per_page))
```

File: app/controllers/admin/api/application_plan_limits_controller.rb [hunk 2]
```diff
@@ -14,15 +14,19 @@ class Admin::Api::ApplicationPlanLimitsController < Admin::Api::BaseController
   ##~ op.description = "Returns the list of all limits associated to an application plan."
   ##~ op.group = "application_plan_limits"
   #
+  ##~ @parameter_per_page_50 = {:name => "per_page", :description => "Number of results per page. Default and max is 50.", :dataType => "int", :paramType => "query", :defaultValue => "50"}
+  #
   ##~ op.parameters.add @parameter_access_token
+  ##~ op.parameters.add @parameter_page
```
Comment (josemigallas) [inline app/controllers/admin/api/application_plan_limits_controller.rb]:
```suggestion
  ##~ op.parameters.add @parameter_current_page
```

File: doc/active_docs/Account Management API.json [hunk 3]
```diff
@@ -1317,6 +1317,20 @@
               "paramType": "query",
               "threescale_name": "access_token"
             },
+            {
+              "name": "page",
```
Comment (josemigallas) [inline doc/active_docs/Account Management API.json]:
```suggestion
              "name": "current_page",
```

File: test/integration/admin/api/application_plan_limits_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    metric = FactoryBot.create(:metric, service: service)
+    metric.usage_limits.create(period: :week, value: 1, plan: app_plan)
+    metric.usage_limits.create(period: :month, value: 1, plan: app_plan)
+
+    get admin_api_application_plan_limits_path(app_plan, format: :json)
+    assert_equal 2, JSON.parse(response.body)['limits'].length
+
+    get admin_api_application_plan_limits_path(app_plan, per_page: 1, format: :json)
+    assert_equal 1, JSON.parse(response.body)['limits'].length
+
```
Comment (josemigallas) [inline test/integration/admin/api/application_plan_limits_controller_test.rb]: Perhaps redundant but it's missing a case:
```ruby
    get admin_api_application_plan_limits_path(app_plan, per_page: 1, page: 2, format: :json)
    assert_equal 1, JSON.parse(response.body)['limits'].length
```

Comment (macejmic) [reply test/integration/admin/api/application_plan_limits_controller_test.rb]: i don't get it. can't see any changes. (?)

Comment (josemigallas) [reply test/integration/admin/api/application_plan_limits_controller_test.rb]: It's page 2, having 1 result, if I'm not mistaken.

Comment (macejmic) [reply test/integration/admin/api/application_plan_limits_controller_test.rb]: it should be per page 1, current page 1, so 1 result.

File: app/controllers/admin/api/application_plan_limits_controller.rb [hunk 5]
```diff
@@ -14,15 +14,19 @@ class Admin::Api::ApplicationPlanLimitsController < Admin::Api::BaseController
   ##~ op.description = "Returns the list of all limits associated to an application plan."
   ##~ op.group = "application_plan_limits"
   #
+  ##~ @parameter_per_page_50 = {:name => "per_page", :description => "Number of results per page. Default and max is 50.", :dataType => "int", :paramType => "query", :defaultValue => "50"}
```
Comment (hallelujah) [inline app/controllers/admin/api/application_plan_limits_controller.rb]: Note to us: https://github.com/3scale/porta/blob/master/app/controllers/admin/api/accounts_controller.rb#L26 is 500

File: app/controllers/admin/api/application_plan_limits_controller.rb (hunk 1 above)
Comment (hallelujah) [inline app/controllers/admin/api/application_plan_limits_controller.rb]: I think a benchmark for this could be good to do. It is performance issue anyway.

Comment (macejmic) [reply app/controllers/admin/api/application_plan_limits_controller.rb]: yep, benchmark sounds good.

File: test/integration/admin/api/application_plan_limits_controller_test.rb (hunk 4 above)
Comment (hallelujah) [reply test/integration/admin/api/application_plan_limits_controller_test.rb]: @macejmic,  @josemigallas meant:

test `1 per page`, and look for `page 2`

Comment (josemigallas) [reply test/integration/admin/api/application_plan_limits_controller_test.rb]: That's right

File: app/controllers/admin/api/application_plan_limits_controller.rb (hunk 1 above)
Comment (macejmic) [reply app/controllers/admin/api/application_plan_limits_controller.rb]:
```
50 usage limits
Admin::Api::ApplicationPlanLimitsControllerTest
Measure Mode: wall_time
Thread ID: 70283998525840
Fiber ID: 70284006297440
Total: 1.915027
```

```
160 usage limits
Measure Mode: wall_time
Thread ID: 70283998525840
Fiber ID: 70284006297440
Total: 5.448363
Sort by: self_time
```

Comment (hallelujah) [reply app/controllers/admin/api/application_plan_limits_controller.rb]: Do you know what is making it slow?
N+1 requests? Missing index or rendering?

Comment (macejmic) [reply app/controllers/admin/api/application_plan_limits_controller.rb]: lots of queriers, i'm on it.

Comment (macejmic) [reply app/controllers/admin/api/application_plan_limits_controller.rb]: it's a bit better.. 

```
50 usage limits
Measure Mode: wall_time
Thread ID: 70289929263520
Fiber ID: 70289961327280
Total: 1.101826
Sort by: self_time
```

**Pull request merged**